add stunnel client implementation

alaypatel07 opened the pull request:
**Describe what this PR does**
This adds stunnel client implementation. Depends on #4 fixes #5 

**Is there anything that requires special attention?**
<!-- Do you have any questions? Did you do something clever? -->

**Related issues:**
<!-- Mention any github issues relevant to this PR -->

alaypatel07 committed:
add stunnel client implementation

Signed-off-by: Alay Patel <alay1431@gmail.com>

JohnStrunk commented on transport/stunnel/client.go:
```diff
@@ -0,0 +1,261 @@
+ client = yes
+ syslog = no
+ output = /dev/stdout
+ [rsync]
```
To match what we did in server?
```suggestion
 [transfer]
```

alaypatel07 replied:
good catch, thanks!

JohnStrunk commented on transport/stunnel/client.go:
```diff
@@ -0,0 +1,261 @@
+
+type client struct {
+	logger         logr.Logger
+	ingressPort    int32
```
I'm struggling a bit w/ the `ingressPort` name. It's `ConnectPort` in the Transport interface, but the `NewClient` params also refer to it as ingressPort. And Endpoint has IngressPort().
Did we make a mistake by having Transport.ConnectPort() and Endpoint.IngressPort()?

alaypatel07 replied:
It's part of a bad refactor, it should be connectPort here in the transport package.

IMO, having ListenPort and ConnectPort at the transport level makes sense because out of the port one is used to accept connection other is used to connect.

In the case of endpoint interface both are used to accept connections, the difference is whether it is exposed external to the cluster or it is internal because of which BackendPort and IngressPort made much more sense, assuming Ingress is the terms used widely to accept traffic into the cluster. Most Kube resources use targetPort and port to mean the same thing respectively but that was turning out to be a little confusing when we were having conversations in the team about this. I would be open to renaming the endpoint interface methods if we want to be consistent with Kube resource fields

JohnStrunk commented on transport/stunnel/client.go:
```diff
@@ -0,0 +1,261 @@
+	return "localhost"
+}
+
+func NewClient(ctx context.Context, c ctrlclient.Client, logger logr.Logger,
```
Please comment the params

JohnStrunk reviewed (commented):
Just a couple little items.
But I am curious about your thoughts on the consistency of the Interface method names. (or maybe having too many days off just left me confused :confused:)

alaypatel07 committed:
fixups to transport/stunnel based on feedback

Co-authored-by: John Strunk <jstrunk@redhat.com>
Signed-off-by: Alay Patel <alay1431@gmail.com>

alaypatel07 commented:
@JohnStrunk Thanks for the feedback, updated a couple of things and added clarification on the naming, PTAL. If we do want to revisit the endpoint interface naming, can we let this PR in and handle it in a separate issue? Thanks

JohnStrunk reviewed (approved):
/lgtm
